### PR TITLE
Recurring Recommendations

### PR DESCRIPTION
**Pull request by LaelWatkins**
The changes / functionality for the Recurring Gifts Recommendation Engine (https://jira.cru.org/browse/IN-62)  is ready for review. 

Insights Tab 
image

Contacts Page - Increase Recommendations Link
image

Contacts Page - Increase Recommendations Contacts
image

### CODE REVIEW

File: app/controllers/contacts_controller.rb [hunk 1]
```diff
@@ -363,4 +363,7 @@ def clear_annoying_redirect_locations
   def contact_params
     @contact_params ||= params.require(:contact).permit(Contact::PERMITTED_ATTRIBUTES)
   end
+
```
Comment (LaelWatkins) [inline app/controllers/contacts_controller.rb]: I never thought/remembered to reply to these comments.  Done.

File: app/controllers/insights_controller.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class InsightsController < ApplicationController
+
+  def index
+    @page_title = _('Insights')
+    @current_user = current_user
```
Comment (LaelWatkins) [inline app/controllers/insights_controller.rb]: This is removed.

File: app/models/obiee.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                  executionOptions:
+                      {async: '',
+                       maxRowsPerPage: -1,
+                       refresh: true,
+                       presentationInfo: true,
+                       type: ''},
+                  sessionID: session_id}
+
+    returned_results = make_call(run_report_client,:executeSQLQuery, run_params)
+    #run_report_client.call( :logoff, message: session_id)  # for some reason this call returns with the expired/invalid session exception
```
Comment (LaelWatkins) [inline app/models/obiee.rb]: This has been deleted.

File: app/views/contacts/_contact.html.erb [hunk 4]
```diff
@@ -40,7 +40,7 @@
       <% end %>
       <% if contact.pledge_amount.to_i > 0 %>
         <%= _('Commitment') %>: <%= contact_exhibit.pledge_as_currency %><br />
-      <% end %>
+      <% end %>z
```
Comment (LaelWatkins) [inline app/views/contacts/_contact.html.erb]: This was deleted.

File: app/views/insights/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+<div class="clr"></div>
+<% if @recommnds %>
+
+<h2>Recurring Gift Recommendations for <%=@current_user %></h2>
+
+<div class="clr"></div>
+<table class="tablelist donations" id="recurringGiftsDataTable">
+  <thead>
+  <tr>
+    <th width="7%">Gift Range</th>
```
Comment (LaelWatkins) [inline app/views/insights/index.html.erb]: All labels have been wrapped in the erb tag and method above.

File: app/views/insights/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+<small><p align="left" position="bottom">+&nbsp;<cite>US Census Bureau. (2012). Retrieved from http://factfinder2.census.gov/ </cite></p></small>
+</p>
+<% else %>
+    <div class="mwrap">
+      <div class="nohide floating" style="background-color: #FDFFD4;">You have no monthly recurring gift recommendations.</div>
+    </div>
+ <% end %>
+
+<script type="text/javascript">
+    $(document).ready(function () {
```
Comment (LaelWatkins) [inline app/views/insights/index.html.erb]: Done.

File: config/routes.rb [hunk 7]
```diff
@@ -161,4 +165,5 @@
 
   # See how all your routes lay out with "rake routes"
 
+
```
Comment (LaelWatkins) [inline config/routes.rb]: Done

File: Gemfile [hunk 8]
```diff
@@ -68,6 +68,8 @@ gem 'rest-client', '~> 1.6.7'
 gem 'retryable-rb', '~> 1.1.0'
 gem 'rollout', '~> 2.0.0'
 gem 'ruby-rtf'
+gem 'savon' , '~> 2.3.0'
+gem 'savon_spec'
```
Comment (LaelWatkins) [inline Gemfile]: Done

File: Gemfile [hunk 9]
```diff
@@ -85,7 +86,7 @@ gem 'whenever', '~> 0.8.1'
 gem 'wicked', '~> 1.0.2'
 gem 'will_paginate', '~> 3.0.3'
 gem 'global_phone', git: 'https://github.com/sstephenson/global_phone.git'
-gem 'global_phone_dbgen'
+#gem 'global_phone_dbgen'
```
Comment (twinge) [inline Gemfile]: Why did this gem get commented out? (though if i'm thinking correctly about what this gem is, it belongs in the :development group)

File: spec/models/insight_analyses_spec.rb [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+  after(:all)  { savon.unmock! }
+
+
+  it 'gets analysis results with designation' do
+
+    creds = {name: APP_CONFIG['obiee_key'], password: APP_CONFIG['obiee_secret'] }
+    fixture = File.read('spec/fixtures/obiee_auth_client.xml')
+
+
+    if savon.expects(:logon).with(message: creds).returns(fixture)
```
Comment (twinge) [inline spec/models/insight_analyses_spec.rb]: Specs should never contain `if` statements. If you explain to me a bit what you're trying to do here, I'll help you figure out the "right" way. 

This comment applies to all the other `if` statements in other specs for this PR as well.

File: spec/models/obiee_spec.rb [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+   0 s_0,
+   "CCCi Transaction Analytics"."- Designation"."Designation Name" s_1
+FROM "CCCi Transaction Analytics"
+WHERE ("- Designation"."Designation Name" = ''Test Desig ((2716653)'')
+ORDER BY 1, 2 ASC NULLS LAST
+FETCH FIRST 10000000 ROWS ONLY'
+
+  obiee = Obiee.new
+
+  it ( 'gets a session id') do
```
Comment (twinge) [inline spec/models/obiee_spec.rb]: Don't use ( ) for `it` statements:

`it 'gets a session id' do `

This comment applies to all the other specs with ( ) as well.

File: Gemfile.lock [hunk 12]
```diff
@@ -27,9 +27,9 @@ GIT
 
 GIT
   remote: https://github.com/CruGlobal/google_contacts_api
-  revision: 49f8b513a4617d763d7488c2c208fb181f9324b4
+  revision: 4aefbb141785763d3f5a0b5fe5bf52ce8b73b99b
   specs:
-    google_contacts_api (0.5.0)
+    google_contacts_api (0.4.0)
```
Comment (twinge) [inline Gemfile.lock]: This gem (and a few others) got downgraded. Was that intentional?

**Comment by twinge** [comment]
You'll need to rebase from master or merge master into this branch and resolve conflicts before we can merge this branch back into master.

File: Gemfile (hunk 9 above)
Comment (LaelWatkins) [inline Gemfile]: It must've been a mistake.  I'm not using this gem. I've sync'd with the master and it's correct now.

File: app/assets/javascripts/insights.js.coffee [hunk 13]
```diff
@@ -0,0 +1,17 @@
+$ ->
+  $().ready ->
+    numberOfrows = document.getElementById("recurringGiftsDataTable").rows.length
```
Comment (twinge) [inline app/assets/javascripts/insights.js.coffee]: You can use jquery to find that $('#recurringGiftsDataTable').size()

**Comment by twinge** [comment]
There are a lot of gem version changes in this PR that makes me nervous about merging this. Can you get rid of all the gem file additions/removals that aren't actually needed by your code?

**Comment by LaelWatkins** [comment]
I've corrected this.  My weirdo Gemfile was the source of my "pain" last
Friday.  I checked out a fresh copy and viola,  all fixed.

On Sat, Jan 31, 2015 at 12:25 PM, Josh Starcher <notifications@github.com>
wrote:

> There are a lot of gem version changes in this PR that makes me nervous
> about merging this. Can you get rid of all the gem file additions/removals
> that aren't actually needed by your code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CruGlobal/mpdx/pull/198#issuecomment-72326921>.
>



-- 
*Lael Watkins*
Senior Programmer Analyst
407 826 2548
lael.watkins@cru.org

File: spec/models/obiee_spec.rb (hunk 11 above)
Comment (LaelWatkins) [inline spec/models/obiee_spec.rb]: This is corrected.  I'm not sure why I had it in this spec.  I didn't have it in any others.

File: spec/models/insight_analyses_spec.rb (hunk 10 above)
Comment (LaelWatkins) [inline spec/models/insight_analyses_spec.rb]: I've corrected this. 

I was trying to make sure the "returns" method wouldn't come back null or incorrect.  I'd forgotten that was part/point of the spec.

What this test does is make stub requests and test them against the responses from the methods in the InsightAnalyses model. 

1. tests analysis results with a desig
2. fails with unknown desig
3. fails with no desig

A stub request is made in each one.

File: Gemfile.lock (hunk 12 above)
Comment (LaelWatkins) [inline Gemfile.lock]: This was definitely a mistake.  I've checked out a fresh copy, commited and pushed it.

**Comment by LaelWatkins** [comment]
I have couple more specs to complete.

File: app/assets/javascripts/insights.js.coffee [hunk 14]
```diff
@@ -0,0 +1,24 @@
+$ ->
+
+  $().ready ->
+  numberOfrows = $('#recurringGiftsDataTable tbody tr').length
```
Comment (twinge) [inline app/assets/javascripts/insights.js.coffee]: Inconsistent case usage in variable name. Should be `numberOfRows`

File: app/assets/javascripts/insights.js.coffee [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+  i=0
+  while numberOfrows > i
```
Comment (twinge) [inline app/assets/javascripts/insights.js.coffee]: This seems like a painful way to loop. I think maybe you can do this (or something similar):

`(num for num in [0..numberOfrows])`

http://coffeescript.org/

**Comment by twinge** [comment]
Made a couple more minor comments, but I'm good with this being deployed to our testers after that.

**Comment by LaelWatkins** [comment]
took me a while :-) but i refactored it to use a comprehension. very nice!

On Mon, Feb 16, 2015 at 2:06 PM, Josh Starcher <notifications@github.com>
wrote:

> Made a couple more minor comments, but I'm good with this being deployed
> to our testers after that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CruGlobal/mpdx/pull/198#issuecomment-74557427>.
>



-- 
*Lael Watkins*
Senior Programmer Analyst
407 826 2548
lael.watkins@cru.org

File: app/assets/javascripts/insights.js.coffee [hunk 16]
```diff
@@ -0,0 +1,19 @@
+$ ->
+
+  $().ready ->
+  numberOfRows = $('#recurringGiftsDataTable tbody tr').length
+  $('#nextAskYearMonth').replaceWith($.datepicker.formatDate 'yy / mm', new Date $('#nextAskYearMonth').text()) for num in [0..numberOfRows]
+
```
Comment (LaelWatkins) [inline app/assets/javascripts/insights.js.coffee]: refactored loop

**Comment by twinge** [comment]
:+1:

File: app/assets/javascripts/angular/controllers/taskShortList.js [hunk 17]
```diff
@@ -3,10 +3,7 @@ angular.module('mpdxApp').controller('taskShortListController', function ($scope
         $scope.tasks = {};
         $scope.comments = {};
         $scope.people = {};
-        $scope.history = page == 'contactHistory';
```
Comment (soberstadt) [inline app/assets/javascripts/angular/controllers/taskShortList.js]: Hi Lael, sorry for not reviewing this closer when you had this open, but was there a reason your removed this stuff from here? It was being used elsewhere.

Comment (LaelWatkins) [inline app/assets/javascripts/angular/controllers/taskShortList.js]: I wouldn't have needed to do so. I shouldn't have touched that js file. I'm
not sure what happened there. Sorry about that. I can add it back in.

Comment (soberstadt) [inline app/assets/javascripts/angular/controllers/taskShortList.js]: OK, thanks for letting me know. I'll do it quick while I'm looking at it now.

**Pull request merged**